service/history/api/describeworkflow/api.go

```diff
@@ -142,12 +142,16 @@ func Invoke(
 				p.LastHeartbeatTime = ai.LastHeartbeatUpdateTime
 				p.HeartbeatDetails = ai.LastHeartbeatDetails
 			}
-			// TODO: move to mutable state instead of loading it from event
-			scheduledEvent, err := mutableState.GetActivityScheduledEvent(ctx, ai.ScheduledEventId)
-			if err != nil {
-				return nil, err
+			if ai.GetActivityType() != nil {
+				p.ActivityType = ai.GetActivityType()
+			} else {
+				// TODO: move to mutable state instead of loading it from event
+				scheduledEvent, err := mutableState.GetActivityScheduledEvent(ctx, ai.ScheduledEventId)
+				if err != nil {
+					return nil, err
+				}
+				p.ActivityType = scheduledEvent.GetActivityTaskScheduledEventAttributes().ActivityType
 			}
```

Comment:
maybe create a method for getting activity type on mutable state? This backward compatibility logic will live for a very long time, so, if later, other places also needs activity type, those places can invoke that method directly.